jujugui/static/gui/src/app/components/user-profile/create-model-button/create-model-button.js

```diff
@@ -0,0 +1,158 @@
+      className: React.PropTypes.string,
+      env: React.PropTypes.object.isRequired,
+      hideConnectingMask: React.PropTypes.func.isRequired,
+      jem: React.PropTypes.object,
+      showConnectingMask: React.PropTypes.func.isRequired,
+      switchModel: React.PropTypes.func.isRequired,
+      user: React.PropTypes.object
+    },
+
+    getInitialState: function() {
+      return {
+        createNewModelActive: false
+      };
+    },
+
+    getDefaultProps: function() {
+      // Just in case broadcastStatus isn't passed in (e.g., in tests), calls
+      // to it should not fail, so default to an empty function.
+      return {
+        className: ''
+      };
+    },
+
+    /**
+      Fetches the model name from the modelName ref and then creates a model
+      then switches to it.
+      @method createAndSwitch
+      @param {Object} e The event handler from a form submission.
+    */
+    createAndSwitch: function(e) {
+      if (e && e.preventDefault) {
+        e.preventDefault();
+      }
+      // The supplied new model name needs to be valid.
+      var modelName = this.refs.modelName;
+      if (!modelName.validate()) {
+        // Only continue if the validation passes.
+        return;
+      }
+      var props = this.props;
+      // Because this will automatically connect to the model lets show
+      // the connecting mask right now.
+      props.showConnectingMask();
+      // XXX This is only for the JIMM flow.
+      props.env.createModel(
+        modelName.getValue(),
+        props.user.user,
+        data => {
+          var err = data.err;
+          if (err) {
+            props.addNotification({
+              title: 'Failed to create new Model',
+              message: err,
+              level: 'error'
+            });
+            console.error(err);
+            props.hideConnectingMask(false);
+            return;
+          }
+          props.switchModel(data.uuid, data.name);
+        });
+    },
+
+    /**
+      Depending on the existance of jem this will either switch to a
+      disconnected model or open up the UI to allow the user to create a
+      new model.
+      @method _nextCreateStep
+    */
+    _nextCreateStep: function() {
+      if (this.props.jem) {
+        // Switch to a disconnected model
+        this.props.switchModel();
+      } else {
+        // Open up the UI to specify a model name for the Controller.
+        this.setState({ createNewModelActive: true }, () => {
+          var input = this.refs.modelName;
+          if (input && input.refs.field) {
+            input.refs.field.focus();
+          }
+        });
+      }
+    },
+
+    /**
+      Generates the elements required for the create new button
+      @method _generateCreateNew
+      @param {String} className The class you'd like to have applied to the
+        container.
+    */
+    render: function() {
+      var classes = classNames(
+        'user-profile__create-new',
+        this.props.className,
```

Comment:
I went back and forth on this. I left it as `this.props.className || ''` for most of the time through the branch, and only changed it over to this at the end. I changed it over because I kept looking for the `getDefaultProps` function, not seeing it, and thinking: "why does this not fall flat on its face"? I like the idea of one standard place for setting default prop values, regardless as to whether there's one or many and they're used in one place or many places.